File: problem-1.java
//TC:O(N*K) where k is the length of longest word
//sc:O(K*26)

class Trie {

    /** Initialize your data structure here. */
    class TrieNode
    {
       boolean isWord;
       TrieNode[] child; 
        
       private TrieNode()
       {
           this.isWord = false;
           this.child = new TrieNode[26]; 
       }
    
        
    }
    
    TrieNode root;

    public Trie()
    {
      root= new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
       TrieNode node =  root; 
       for(int i=0;i<word.length();i++)
       {
         char ch = word.charAt(i);
         if(node.child[ch-'a']==null)
             node.child[ch-'a']=new TrieNode();
           
         node =  node.child[ch-'a']; 
       }
        
      node.isWord=true;  
        
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) 
    {
        TrieNode node = root;
        
        for(int i=0;i<word.length();i++)
        {
            char ch = word.charAt(i);
            if(node.child[ch-'a']==null)
                return false;
            
            node = node.child[ch-'a'];
        }
        
        return node.isWord;
        
        
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix)
    {
        TrieNode node = root;
        
        for(int i=0;i<prefix.length();i++)
        {
           char ch = prefix.charAt(i); 
           if(node.child[ch-'a']==null)
               return false; 
            node=node.child[ch-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: problem1.cpp
// Time Complexity : search insert and starts with O(l) // length of the string
// Space Complexity : O(l) length of the string
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

#include<iostream>
#include<vector>
#include<string>
using namespace std;

class Trie {
    
public:
    vector<Trie*> child;
    bool flag;
    Trie() {
        this->child.assign(26,NULL);
        this->flag = false;
    }
    
    void insert(string word) {
        Trie* obj{this};
        int n = word.length();
        for(int i{};i<n;++i){
            int idx = word.at(i) - 'a';
            if(obj->child.at(idx) == NULL){
                obj->child.at(idx) = new Trie();
            }
            obj = obj->child.at(idx);
        }
        obj->flag = true;
    }
    
    bool search(string word) {
        int n = word.length();
        Trie* obj{this};
        for(int i{};i<n;++i){
            int idx = word.at(i) - 'a';
            obj = obj->child.at(idx);
            if(obj == NULL) return false;
        }
        return obj->flag;
    }
    
    bool startsWith(string word) {
        int n = word.length();
        Trie* obj{this};
        for(int i{};i<n;++i){
            int idx = word.at(i) - 'a';
            obj = obj->child.at(idx);
            if(obj == NULL) return false;
        }
        return true;
    }
};